js/codemeta_generation.js

```diff
@@ -115,49 +128,65 @@ function generatePersons(prefix) {
     return persons;
 }
 
-function generateCodemeta() {
-    var inputForm = document.querySelector('#inputForm');
-    var codemetaText, errorHTML;
 
-    if (inputForm.checkValidity()) {
-        var doc = {
-            "@context": "https://doi.org/10.5063/schema/codemeta-2.0",
-            "@type": "SoftwareSourceCode",
-        };
-
-        let licenses = getLicenses();
-        if (licenses.length > 0) {
-            doc["license"] = (licenses.length === 1) ? licenses[0] : licenses;
-        }
+function buildDoc() {
+    var doc = {
+        "@context": CODEMETA_CONTEXT_URL,
+        "@type": "SoftwareSourceCode",
+    };
 
-        // Generate most fields
-        directCodemetaFields.forEach(function (item, index) {
-            doc[item] = getIfSet('#' + item)
-        });
+    let licenses = getLicenses();
+    if (licenses.length > 0) {
+        doc["license"] = (licenses.length === 1) ? licenses[0] : licenses;
+    }
 
-        doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
+    // Generate most fields
+    directCodemetaFields.forEach(function (item, index) {
+        doc[item] = getIfSet('#' + item)
+    });
 
-        // Generate simple fields parsed simply by splitting
-        splittedCodemetaFields.forEach(function (item, index) {
-            const id = item[0];
-            const separator = item[1];
-            const value = getIfSet('#' + id);
-            if (value !== undefined) {
-                doc[id] = value.split(separator).map(trimSpaces);
-            }
-        });
+    doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
 
-        // Generate dynamic fields
-        var authors = generatePersons('author');
-        if (authors.length > 0) {
-            doc["author"] = authors;
-        }
-        var contributors = generatePersons('contributor');
-        if (contributors.length > 0) {
-            doc["contributor"] = contributors;
+    // Generate simple fields parsed simply by splitting
+    splittedCodemetaFields.forEach(function (item, index) {
+        const id = item[0];
+        const separator = item[1];
+        const value = getIfSet('#' + id);
+        if (value !== undefined) {
+            doc[id] = value.split(separator).map(trimSpaces);
         }
+    });
 
-        codemetaText = JSON.stringify(doc, null, 4);
+    // Generate dynamic fields
+    var authors = generatePersons('author');
+    if (authors.length > 0) {
+        doc["author"] = authors;
+    }
+    var contributors = generatePersons('contributor');
+    if (contributors.length > 0) {
+        doc["contributor"] = contributors;
+    }
+    return doc;
+}
+
+// FIXME this hack to make expansion work for now
+// (expansion does not accept empty values)
+function deleteEmptyValues(doc) {
+  return JSON.parse(JSON.stringify(doc));
+}
```

Comment:
what do you mean by "empty values"?

Reply:
The way `doc` is currently built, it adds properties (like `@id`) whether the value is empty or not, which breaks the expansion (with the error `"@id" value must a string`). I added this fix to make it work without overloading the PR but it is something to fix by directly updating the way `doc` is built (IMHO).

Reply:
I think I managed to fix it with:

```diff
--- a/js/codemeta_generation.js
+++ b/js/codemeta_generation.js
@@ -107,7 +107,10 @@ function generateShortOrg(fieldName) {
 function generatePerson(idPrefix) {
     var doc = {
         "@type": "Person",
-        "@id": getIfSet(`#${idPrefix}_id`),
+    }
+    var id = getIfSet(`#${idPrefix}_id`);
+    if (id !== undefined) {
+        doc["@id"] = id;
     }
     directPersonCodemetaFields.forEach(function (item, index) {
         doc[item] = getIfSet(`#${idPrefix}_${item}`);
```